### ISSUE TITLE
Conventions

### ISSUE THREAD

**Issue by nfeske** [cbe-project/cbe#1]
Hereby, I'd like to propose common conventions for the design and implementation. The proposal is the result of our practical experimentation with SPARK and by our early brain storming about the CBE design.

# Design methodology

We will conduct the rapid design-space exploration using our mother tongue (C++) while using only language features that can be 1:1 mapped to SPARK concepts. This way, we can efficiently validate our individual decisions by building intermediate programs that can be executed and stressed separately. Since these programs are Genode components, they are hybrids of C++ and SPARK. C++ is used for the glue to the outside Genode system whereas SPARK is used for the actual domain logic (aka the "meat") that is solely expressed statically.

By viewing C++ and SPARK as a continuum rather than an either-or-decision, we are able to iterate quickly, and test each part of the CBE implementation individually before attaining the complete picture.

We deliberately constrain the control flow between C++ and SPARK as follows:

* C++ code calls C++ code.
* C++ code calls SPARK code.
* SPARK code calls SPARK code.
* **SPARK code never calls C++ code!**

This implies that SPARK code implements state machines but cannot block for I/O. From C++ perspective, SPARK code is treated like a self-contained library.

The ABI between C++ and SPARK code must use bit-precise types.

* Method arguments must be unambiguously based on `uint32_t` and such.
* If the ABI uses enum values, we have to use `enum class`, specifying the underlying storage type.
* If the ABI uses structured types as arguments (C++ struct / SPARK record), the C++ type must use `__attribute__((packed))` and the SPARK record should be defined via the `with Size => <NUMBER-OF-BITS>` annotation.

# Modular design

In https://github.com/genodelabs/genode/commit/98182379188fab9f262a8df8623d9fa4843a0ddc I present a way to seamlessly integrate SPARK-implemented objects into C++ programs. Such objects can be used in the same way as C++ objects. E.g, they can be instantiated any number of times, allocated statically (BSS), or dynamically, or used as member variables, or as a stack variable.

*Note that there is one caveat still, which is the dimension of the opaque `_space` of the C++ object, which must always correspond to the size of the corresponding SPARK record. To avoid the need for manually maintaining this consistence, we may consider a checking tool in the future, but not now.*

This approach allows us to gradually "SPARKify" Genode components by successively replacing C++ objects with SPARK objects. To support this notion, let us constrain the SPARK code as follows:

## All packages are pure

There are no package variables. State is kept in records only. Records are instantiated explicitly, from the outside of the package.

This guarantees us that the CBE code is free from any side effects. The allocation of memory is solely defined outside the CBE packages. In a Genode component, this would be the C++ code. In a raw Ada/SPARK program, this would be some form of a (non-pure) "Main" package where the top-level code instantiates records of the CBE packages as package variables.

As another benefit of this approach, there is no notion of any singleton. E.g., since the instantiation of each part of the CBE is at the discretion of the CBE-using code, a Genode component that serves multiple client sessions can host an instance of the CBE for each session.

## Representation of a simple C++ class in SPARK

In SPARK, we model the concept of a C++ class as follows. Let's use an example class `Machinery` (as I used in the commit cited above).

* There is a dedicated package called `Machinery`.

* The member variables are stored in a private record type suffixed with `_type`. E.g., `Machinery_Type`.

* The constructor is a procedure `Initialize` expecting the object as out parameter:

  ```
  procedure Initialize (Machinery : out Machinery_Type)
  ```

Our notion of a class does not support inheritance or virtual functions. It is just an opaque object with a method interface to interact with the object.

**Comment by m-stein** [comment]
I totally agree with what you suggest and have only one small addition. Naming the package state (or object) like the package may be confusing in larger projects and may even lead to compile-relevant ambiguities like in the following code snippet:
```
with Inner;
package Outer is
  type Outer_Type is record
    Inner : Inner.Inner_Type; -- Error
  end record;
  function Func(Inner : Inner.Inner_Type) return Something is -- Error
  begin                                                                        
    return Inner.Func(Inner); -- Error
  end Func;
end Outer;
```
Furthermore, I think we can use a more generic name for the state type, like `Inner.Object_Type`/ `Inner.State_Type` and name the corresponding states/objects `Inner_Object`/`Inner_State` (outside the package) respectively `Object`/`State` (inside the package).

**Comment by nfeske** [comment]
Are you sure that the errors don't originate from naming the record for the `Outer` package `Outer` instead of `Outer_Type`?

Your suggestion sounds good to me. I agree that repeating the package name is not very appealing. I would avoid the term `State` though. A type should better be called a type. ;-) To take your suggestion one step further, if we follow the convention to have only one "class" type defined per package, we could simply call it `Type`. E.g., the `Outside` package would speak of `Inner.Type`. That conciseness would be quite nice.

That said, we should follow the advise of seasoned SPARK developers here. @senier, what do you think?

**Comment by m-stein** [comment]
Sorry, in my (more complex) example, the record type had a name that differed from the package name.  I've adapted the simplified example.

**Comment by senier** [comment]
@m-stein: The problem in your example is that you call the record component `Inner`, like the package you're withing. (Also, your spec file implements a function body, which must go into the body part ob be refactored into an expression function). This would work:

```Ada
with Inner;
package Outer
is
  type Outer_Type is record
    I : Inner.Inner_Type;
  end record;
  function Func (I : Inner.Inner_Type) return Inner.Something is (Inner.Func(I));
end Outer;
```

```
package Inner
is
   type Inner_Type is new Integer;
   type Something is new Integer;
   function Func(Inner : Inner_Type) return Something;
end Inner;
```

@nfeske: I agree that repeating the package name is inconvenient and unnecessary. Your cannot name your entity `type`, as it is a reserved word. In your binding generator we use `Class` as a type name for bound classes. For referring to a class we then yous `Class_Package.Class`. Pretty similar to what you suggested. I normally prefer to add a `_Type` suffix to custom types, but in this particular case we found a simpler name more convenient.

**Comment by nfeske** [comment]
@senier Thanks for the quick response. `Class` is indeed very fitting. We can still follow the `_Type` convention for non-class types.

**Comment by senier** [comment]
Some more remarks concerning @nfeske's initial suggestion (thanks for that!):

1. As the CBE core is not only to be used from C++, an abstraction between the types in the C++-Interface and the SPARK core library would be beneficial. While necessary for interfacing, you don't want to have modular types in your SPARK code most of the time (they wrap and tend to make life much harder during prove). Also, the use of `'Size` should be limited as much as possible as calculating the right sizes is cumbersome and may be a maintenance issue. When don right, it's not strictly necessary for interfacing with C++ either (see below).

2. I have issues with the way C++ classes are imported in your example (Convention "C" and C++ names). While it may work now, I'm not sure what actually is guaranteed by the language/compiler. The correct way for importing C++ classes is through a limited record type using `Convention => CPP` and importing the constructor using `pragma CPP_Constructor`. There are a couple of examples in [our binding generator](https://github.com/Componolit/Cappulada/blob/issue_52/tests/data/test_class_with_functions.ads). IMO we should follow the "official" way of binding C++ code to avoid surprises further down the road.

3. I second your suggestion to make packages pure and stateless. Unfortunately this may not always be easily possible. Pure packages must not with non-Pure packages and some system packages are not pure. In libsparkcrypto I recently tried to purify all packages, but due to the use of GNAT.IO for debug ouput (which is `Preelabore`, but not `Pure`) I ended up with some package being preelaborate and others being pure. Removing debug output may not really be an option either ... we need to see how well that works out. We should aim at pure and at least make them preelaborate (especially as we do not execute elaboration code at the moment!).

**Comment by m-stein** [comment]
@senier That's exactly the problem I wanted to point out in the inital suggestion of @nfeske but I aimed for an - as I think - more descriptive solution. Instead of naming the package object `I` outside the package and `Inner` inside, I would suggest to consistently suffix it with something like `_Object`. This would also pretty directly reflect the overall design approach of "classes" that are composed of a member record (the object) and a package (the methods)

**Comment by nfeske** [comment]
Thank you @senier for the very useful remarks!

1. Yes, we should better leave the CBE package untainted with any C++-isms. It is probably the best to keep the C++ interfacing code (C++ symbol names, `with Size` annotations) in a wrapper package around the CBE package. This wrapper package could then also convert the "unclean" modular types as used towards C++ to the semantically strict types used within the CBE. With this approach, we could actually consider turning the `Block_Size` into a generic argument (CBE would be a generic), instead of a constant because the instantiation of the CBE would happen solely in the wrapper package. But maybe this goes too far?

2. Thanks for the remarks and the pointer. @m-stein Would you like to update the example in Genode's ada test accordingly?

3. The tainting of the pureness by the debug messages is really unfortunate. I wonder, isn't there any way around that? After all, the "side effects" of the printed messages are completely irrelevant. Maybe debug messages could not be printed with the overly broad GNAT.IO in the first place? Would it be possible to define a custom "pure" interface solely designated for printing debug messages and let this interface call an external back-end `Write_Debug_Message` function provided by a raw C binding? GNAT wouldn't know what happens behind this binding stay happy.

**Comment by senier** [comment]
@m-stein: It seem like I should clarify what C++ entities we map to which entities in Ada. Admittedly this can be a bit confusing.

1. C++ class declarations are mapped to Ada record types. As Ada records only have data members, only members of a C++ class are considered (types and order, of cause, have to match).
2. C++ class members are mapped to Ada functions/procedures (depending on whether this is a void method or not) that have a pointer (or in/out parameter) to the above record type.
3. Where the record declaration and the declaration of the member functions is done, does not really matter. Package specs in Ada are just collection of types and subprogram declarations, roughly comparable to namespaces in C++. Conventionally, we name the package hierarchy after the Namespace/Class hierarchy in C++ (so. `Foo::Bar::Baz::My_Class` would become `Foo.Bar.Baz.My_Class.Class`, where Class is the record type mentioned above and Class methods are declared in the `Foo.Bar.Baz.My_Class` package)
4. When creating an instance of a C++ class  in Ada, you just create a object of the record type initialized by its imported constructor (which also is just a subprogram with the special CPP_Constructor pragma).

As for special naming rules - we normally try to stay as close to the C++ interface as possible so you don't have to create additional documentation for an Ada binding. Of cause that fails where the C++ implementation uses reserved words (e.g. type ;-) or invalid names (e.g. leading or double underscores). In this case we perform some systematic escaping. For entities without a counterpart in C++ we use descriptive names (Constructor for constructors and Class for the record type matching the class definition).

Do you see any benefit in special naming rules beyond that? For objects I would normally chose something descriptive without any special pattern, e.g.

```
Block_Interface : CBE.Genode.Block.Class := CBE.Genode.Block.Constructor (42)
```

[Here](https://gist.github.com/senier/d3a7f3ffdee8fdcd26e7456b43d04206) is a complete example of how to implement a class method in Ada while the class is defined in C++ (generated automatically using Cappulada, except for the `Import` of the method changed to `Export`). I believe this is the most official way of doing that and we should adopt if for the interfaces Genode <-> CBE. This way we may avoid the class size issue.

**Comment by senier** [comment]
@nfeske:

1. Making the CBE package generic should be possible, but I would start with a non-generic implementation. Generic packages tend to be less convenient to prove and I can't foresee what implication it has on the whole C++/Ada interfacing topic. Making packages generic after the fact is typically manageable in Ada.

3. You're right - making the packages impure just for debugging sounds like a bad deal. I reopened [the respective issue](https://github.com/Componolit/libsparkcrypto/issues/8) in libsparkcrypto to see whether there is a hack for debug output in pure packages. One obstacle may be the implementation permission in the Ada standard to remove/cache subprogram calls to a pure package, as they can never have a different result with the same arguments. We need to see what GNAT actually does here.

**Comment by senier** [comment]
NB: It is not necessary to use `enum class`, the important thing is the type specifier. Something like:

```
enum Foo : int { Elem1, Elem2, Elem3 };
```

is sufficient. When generating the Ada spec with Cappulada, we set the size of the Ada enumeration type to the size of the underlying type of a C++ enum. This also works if no explicit type specifier is present for the C++ enum. However, if I understand the C++ rules correctly, the compiler may chose a base type that fits all the enum value (at most the size of `int`), unless it is forced to use a bigger type. Consequently, Clang may choose `unsigned char` for an enum with values 0-2 and GCC may choose `int` for the same enum.

Using an explicit type is certainly safest here.

**Comment by nfeske** [comment]
That's a very insightful discussion. It's nice that we have it on record.

@senier please note that we are not using C++ classes from Ada code within the CBE. Do you deem it necessary to introduce one convention that fits both directions (the use of C++ from Ada and also the use of SPARK from C++)? I'm a bit afraid of introducing a level of bureaucracy that feels unnatural on either of both sides.

To illustrate my concern, in issue #2 we aimed at the separation of the C++-ABI-specific SPARK code from the actual CBE code using separate wrapper packages. This way, the SPARK code can benefit from stricter types and remains untainted by C++-isms that are unimportant for the CBE's actual semantics (like the mangled symbol names, or the looser numeric types that correspond to C++ enums).

To me, it would feel natural to host those wrappers within a sub package `CBE.Cxx`. Each type that is used at the CBE's ABI would have a representation there, e.g., `CBE.Cxx.Request`, `CBE.Cxx.Splitter`. When the CBE is used from a Ada-only component, the `CBE.Cxx` package (and sub packages) can just be ignored. The types of those ABI sub packages would map to the types `CBE.Request`, `CBE.Splitter`, which would be idiomatic SPARK types (as strict as they can get). Now, on the C++ side, we want to call those classes `Cbe::Request`, `CBE::Splitter`, not `Cbe::Cxx::Request`, `Cbe::Cxx::Splitter`.  By enforcing a strict convention about the mapping of C++ namespaces to packages and vice versa, I'm afraid we lose this flexibility.

I think it is fantastic that Ada gives us the freedom to map an Ada symbol to any C++ symbol we wish. This is very powerful. Let's use this power to our advantage to attain code that looks perfectly natural on each side of the fence.

As a minor question regarding the `Class` convention, I'm wondering, why do you propose suffixing a package that contains a `Class` record with `_Class`? Wouldn't it be nicer to speak of `Foo.Bar.Baz.My.Class` instead of `Foo.Bar.Baz.My_Class.Class`?

> It is not necessary to use enum class, the important thing is the type specifier.

Thanks for this remark. I wasn't aware that the type specifier can be used on plain `enum`. I thought this feature is available for enum classes only.

**Comment by senier** [comment]
> To me, it would feel natural to host those wrappers within a sub package `CBE.Cxx`. Each type that is used at the CBE's ABI would have a representation there, e.g., `CBE.Cxx.Request`, `CBE.Cxx.Splitter`. When the CBE is used from a Ada-only component, the `CBE.Cxx` package (and sub packages) can just be ignored. The types of those ABI sub packages would map to the types `CBE.Request`, `CBE.Splitter`, which would be idiomatic SPARK types (as strict as they can get). Now, on the C++ side, we want to call those classes `Cbe::Request`, `CBE::Splitter`, not `Cbe::Cxx::Request`, `Cbe::Cxx::Splitter`. By enforcing a strict convention about the mapping of C++ namespaces to packages and vice versa, I'm afraid we lose this flexibility.

I like that approach. Admittedly, I didn't recognize that issue if we map namespaces to packages directly. It's not problem, of cause, to root the packages in a different hierarchy.

>  Let's use this power to our advantage to attain code that looks perfectly natural on each side of the fence.

Agreed.

>  As a minor question regarding the `Class` convention, I'm wondering, why do you propose suffixing a package that contains a `Class` record with `_Class`? Wouldn't it be nicer to speak of `Foo.Bar.Baz.My.Class` instead of `Foo.Bar.Baz.My_Class.Class`?

Yep, the `_Class` suffix was a mistake.

**Comment by m-stein** [comment]
@senier 
1. Regarding your points 1.-3.: I see your point that it is not _neccessary_ that a C++ class is translated to the constellation of an Ada-package that contains appropriate subprograms for the function members and an appropriate private record for the data members. But I intentionally applied this restriction as I assume it to be pretty handy.

2. A question that maybe was answered already offline but whose answer I seem to have missed is:
Which role plays Cappulada or automatic binding generation in general in this project?

3. Regarding your point 4.: If I understand the so far depicted design methodology right, it should never be necessary to call a C++ constructor from Ada but always the other way around an Ada subprogram that initializes a record (for which only a placeholder struct with the public interface is declared in C++). 

@nfeske @senier 
> Would you like to update the example in Genode's ada test accordingly?
4. Yes, but there are some issues that gave me the impression that we should rethink this undertaking:
* `pragma CPP_Constructor` is only for _importing_ constructors from C++, but in the example, we want to _export_ a "constructor" from Ada. It is not allowed to define the implementation of a `CPP_constructor` marked function in Ada. [[GNAT RM]](https://gcc.gnu.org/onlinedocs/gnat_rm/Pragma-CPP_005fConstructor.html)
* While I see that limited records are nice for importing C++-implemented classes (so the C++ operators are not silently by-passed by the Ada standard operators) it leaves me a bit puzzled why to make a record limited that is in C++ merely reflected through a placeholder with the public interface declaration (all the function/data member definitions are in Ada anyway). [[WikiBook]](https://en.wikibooks.org/wiki/Ada_Programming/Types/limited)
* Marking the record `Import, Convention => CPP` seems contradictory to me and requires it to be limited. We could rather mark it `Export, Convention => CPP` but I don't know whether this has any benefit for us as I haven't found any helpful documentation so far. Maybe it allows us to bind the declaration of the C++ placeholder struct more thightly to the Ada size/subprograms?

**Comment by senier** [comment]
This multi-dimensional ticket starts getting hard to follow ;-)

@m-stein: You're right that we never call a C++ constructor from Ada. Hence, the `CPP_Constructor` may seem counter-intuitive. However, GNAT insists on having a constructor imported when importing a record as C++ class.

[Here](https://gist.github.com/senier/d3a7f3ffdee8fdcd26e7456b43d04206) is an example that IMHO matches our use case pretty directly: We define a C++ class, import it as Ada record type. In the C++ main function we call the constructor and invoke a class method that is implemented in Ada. Isn't that what we want? The C++ code looks like ordinary C++ code (no class size hacks or similar) and the Ada code looks pretty standard too.

We develop Cappulada to create complex C++ binding with the goal to have an Ada-only API to Genode. It does not play a role in the development of CBE's core, but we will use it for the tests. As we have it anyway, the translation from the C++ class to an Ada package could be automated.

I do not insist on that approach, though. I'm just not comfortable with that uncommon way of C++ interfacing, as the compiler does things behind the scenes we may not be aware of.

**Comment by senier** [comment]
Libsparkcrypto is fully pure now (not yet merged to the main branch). [Here](https://github.com/Componolit/libsparkcrypto/issues/8) is a description how the hack works. We should employ the same strategy for CBE in cases where printf-style debugging is needed.

**Comment by m-stein** [comment]
I have a short suggestion to make:
I'm currently postfixing all type names with ``_Type`` to prevent name clashes with package and instance names. However, several times during the last weeks I felt the desire to replace ``_Type`` with the shorter ``_T`` to make code more readable. Would this be fine for all involved parties?

**Comment by nfeske** [comment]
I'm impartial regarding the use of `_Type` versus `_T`. As long as the convention is applied consistently, I'm fine.

**Comment by senier** [comment]
@m-stein, I see your point, but honestly, I dislike the `_T` suggestion as it deviates from the Ada way of naming things (no abbreviation if possible). Out of curiosity, I had a look at the [respective section of the Ada style guide](https://en.wikibooks.org/wiki/Ada_Style_Guide/Readability#Names) and had to learn that they do not follow the `_Type` approach either. Seems like they're mandating self-explanatory names which have no further structure. I'd suggest we look through some of the types you already have in you're code tomorrow to see whether omitting the suffix is a viable option (or just makes the code incomprehensible).

**Comment by nfeske** [comment]
This topic has been settled long ago.

**Comment by m-stein** [comment]
@nfeske Thanks, I hoped that you would consider these issues too :)
